Support for \raisebox

edemaine opened the pull request:
I started working on `\raisebox` support which will help for #657 and #224.  Currently `\raisebox{...ex}{...text...}` "works", except that the bounding box of the resulting typeset text isn't set correctly, so raised/lowered objects cut into the lines before/after.  I guess I'm still not fully up to speed on how HTML generation works here -- based on other code in `buildHTML.js`, I tried manually updating `.height` or `.depth` according to whether the box moved up or down, but that didn't work.  If anyone has a chance to look at this, I'd appreciate help fixing it.

Here's a sample render with the test server:

![raisebox](https://cloud.githubusercontent.com/assets/2218736/25505530/23974ddc-2b70-11e7-804e-ff23465f9545.png)

Here's the same example rendered in TeX, so seems good except for the bounding box issue:

![raisebox_tex](https://cloud.githubusercontent.com/assets/2218736/25505565/523a508a-2b70-11e7-843c-44db2bde732d.png)

Also fixed a typo where parsed sizes are accidentally labeled as "color" nodes instead of "size" nodes.

edemaine commented on src/buildHTML.js:
```diff
@@ -1491,6 +1491,22 @@ groupTypes.mclass = function(group, options) {
+            span.height += calculateSize(dy.number, options.style);
+        } else {
+            span.depth -= calculateSize(dy.number, options.style);
+        }
```
The idea is that the height increases if the box is moved up, and the depth increases if the box is moved down.  At least, my mental model is that height is the extent above the baseline (y-max) while depth is the extent below the baseline (y-min) -- please correct me if I'm wrong.  However, this code doesn't seem to be working, and I'm not really sure how `height` and `depth` end up getting used (or not).

kevinbarabash replied:
> At least, my mental model is that height is the extent above the baseline (y-max) while depth is the extent below the baseline (y-min) -- please correct me if I'm wrong.

That's right, but if a glyph is shifted vertically, the the extent below the baseline will be different.

What happens if you put the test expression inside a `\sqrt` or a `\frac`?  It may be that just our wrapper container isn't doing the right thing with the `height` and `depth` you're setting on the `span`.

ronkok replied:
Inside a `\frac` or `\sqrt`, the bounding box looks good:
![raisebox](https://cloud.githubusercontent.com/assets/16403058/26038466/ccaa188c-38bd-11e7-8299-912fc01cb6d6.PNG)

This screenshot, by the way, was done with code that was slightly tweaked:

```
groupTypes.transform = function(group, options) {
    const span = groupTypes.text(group, options);
    const dy = calculateSize(group.value.dy.value, options.style);
    if (dy > 0) {
        span.height += dy;
    } else if (dy < 0) {
        span.depth -= dy;
    }
    span.style.top = -dy + "em";
    span.style.position = "relative";
    return span;
};
```

kevinbarabash replied:
@ronkok did your tweaks have any effect on the top level bounding box?

ronkok replied:
Now that you ask me, yes they did. Odd. I was only trying to clean it up, not change how it worked. Here's a screenshot from the original code:
![raisebox2](https://cloud.githubusercontent.com/assets/16403058/26038792/2c736122-38c5-11e7-8739-b459c4f37ffb.PNG)
Now that I need to take this more seriously, here’s a version that restores the outer if statement.

```
groupTypes.transform = function(group, options) {
    const span = groupTypes.text(group, options);
    if (group.value.dy) {
        const dy = calculateSize(group.value.dy.value, options.style);
        if (dy > 0) {
            span.height += dy;
        } else if (dy < 0) {
            span.depth -= dy;
        }
        span.style.top = -dy + "em";
        span.style.position = "relative";
    }
    return span;
};
```
That code gives me the good results.

edemaine commented on src/buildHTML.js:
```diff
@@ -1491,6 +1491,22 @@ groupTypes.mclass = function(group, options) {
+    const span = groupTypes.text(group, options);
+    if (group.value.dy) {
+        const dy = group.value.dy.value;
+        span.style.top = "" + (-dy.number) + dy.unit;
```
Good point -- indeed, it currently does not, but would be easy to tweak to.  (A bigger issue is that `calculateSize` only supports mu and em.)

kevinbarabash replied:
It looks like `calculateSize` supports `ex`.
```
const calculateSize = function(sizeValue, style) {
    let x = sizeValue.number;
    if (sizeValue.unit === "ex") {
        x *= style.metrics.emPerEx;
    } else if (sizeValue.unit === "mu") {
        x /= 18;
    }
    return x;
};
```

edemaine commented:
Awesome, thanks @kevinbarabash and @ronkok for finding the solution here!  Looks like it was close indeed, and switching to `calculateSize` at the top level was the missing link.

Here is what my original example `A\raisebox{-0.5em}{B}\raisebox{0.5em}{C}D` looks like with the new code (now pushed):

![abcd](https://cloud.githubusercontent.com/assets/2218736/26083990/df966d8a-39a6-11e7-91f1-b8b7c68aa16c.png)

It's better, though the top margin is still a lot smaller (zero or even slightly less) than a straight `\text{C}`:

![c](https://cloud.githubusercontent.com/assets/2218736/26084005/047be206-39a7-11e7-8e50-6d498b78f92c.png)

ronkok commented:
You mentioned a smaller margin that a straight `\text{C}`. I wonder if that is why buildCommon.makeVList spends the effort to call makeFontSizer.

edemaine commented:
Hey, I got it!  I was looking at the definition of the bottom strut in `buildHTML.js`, and it uses a `vertical-align` to put the bottom strut below the top strut.  I didn't know you could put numbers into `vertical-align`.  So I tried that instead of a `top` style (see [this trivial diff](https://github.com/Khan/KaTeX/pull/685/commits/da322228a220256b73eccb2694d66574b5102915)), and it now has exactly the right bounding box!

![abcd](https://cloud.githubusercontent.com/assets/2218736/26084316/42385ea6-39a9-11e7-8272-2a1b0a702090.png)

kevinbarabash commented:
> I didn't know you could put numbers into vertical-align

Neither did I.  Cool!

kevinbarabash commented:
@edemaine now that everything's working correctly, can you add a screenshot test?

edemaine commented:
In adding a screenshot test similar to @ronkok's, I realize that we're not quite there yet: `vertical-align` behaves well at the top level (and in `\frac`s) but doesn't seem to interact properly with `\sqrt`...

![abc](https://cloud.githubusercontent.com/assets/2218736/26086423/7d4801cc-39b8-11e7-8c05-6375bdb0dac0.png)

It seems that the positive `vertical-align` in the b is causing the ab to move down... weird...  Hmm, I think this is related to `makeFontSizer` that @ronkok mentioned, as the contents of the sqrt are put in a `vlist` (although so are the `\frac`s...).

ronkok commented:
I don't think the problem is yours. If I understand matters correctly, KaTeX does not change the radical height in a continuous manner. It's a step function. You happened to land on a bad height.

edemaine commented:
@ronkok Here's what it looks like, switching back to `top` setting:

![abc](https://cloud.githubusercontent.com/assets/2218736/26086924/b2e78ae8-39bb-11e7-86e6-3041e4f8e861.png)

The only difference is the vertical shift in the `a\raisebox{0.5em}{b}`.

I believe the issue is [Movement Of the Line Box’s Baseline](http://christopheraue.net/2014/03/05/vertical-align/#movement-of-the-line-boxs-baseline) (quirks of vertical-align).  Still grokking...

edemaine commented:
Upon further investigation, I think the `top` version (essentially @ronkok's version) was actually correct.  There is a larger space at the top of `\text{a}` simply because of the overall line height.  This margin disappears even in a simple example like `a^{a^{a^a}}`:

![aaaa](https://cloud.githubusercontent.com/assets/2218736/26087569/7d1b06ac-39bf-11e7-94e0-f99278ac434f.png)

So I've reverted to that version, and added a screenshot test.  I think we're back to good-to-go.

kevinbarabash commented:
@edemaine you may want to the texcmp docker to compare the output against LaTeX's output.

edemaine commented:
One more question: I currently call the parse node "transform", with the thought that it could be used for rotatebox and such in the future. But I wonder if just calling it "raisebox" and keeping it simple would be better. Let me know your opinion.

kevinbarabash commented:
Good question.  I think having separate `raisebox` and `rotatebox` methods probably makes since we'll need to do something pretty different to implement `rotatebox`.

edemaine commented:
@kevinbarabash Good call on using texcmp.  I see now that LaTeX resets font size to full size when switching to text mode via `\raisebox` (which `\text` probably is careful to avoid), and I need to do the same:

![raisebox](https://cloud.githubusercontent.com/assets/2218736/26107842/b08dc480-3a18-11e7-9ba4-1f80d3671485.png)

According to [this ImageMagick page](http://www.imagemagick.org/Usage/color_mods/#grayscale), these are probably equivalent, so perhaps we could change that to support more ImageMagick versions?

edemaine commented:
In principle I fixed it, but I've run into a bug: #709

kevinbarabash commented:
@edemaine I've merged the fix for #709.

edemaine commented:
I rebased.  The switch to `\normalsize` is working now, but the bounding box computation seems completely broken (unphased by the `\raisebox`).  In this texcmp, the red (KaTeX) raised b's intersect whatever is above them (`\frac` line or `\sqrt` line).

![raisebox](https://user-images.githubusercontent.com/2218736/27747864-296a234e-5d9b-11e7-95dc-48184ce46b0a.png)

edemaine commented:
I got another chance to look this over and revise substantially.  I believe this is now working almost perfectly.  I use `makeVList` to do the vertical shift, so it should be well-behaved (and it appears at least reasonably so).  Internal to buildHTML, I use `sizing` and `text` nodes to switch to the equivalent of `\textrm{\normalsize...}`.  texcmp is almost perfect (slight horizontal spacing difference?):

![raisebox](https://user-images.githubusercontent.com/2218736/30039232-36daeed4-919c-11e7-8216-26be6fdeb71e.png)


I also slightly cleaned up the parse tree: `text` nodes were using a `style` attribute to specify a font, but styles mean something else in KaTeX; switched the attribute to `font`.

Probably best to look at this [as a complete diff](https://github.com/Khan/KaTeX/pull/685/files) but if you prefer you can look at just the last commit for the changes relative to previous versions.

kevinbarabash commented on src/buildHTML.js:
```diff
@@ -304,7 +304,7 @@ groupTypes.ordgroup = function(group, options) {
 
 groupTypes.text = function(group, options) {
-    const newOptions = options.withFont(group.value.style);
+    const newOptions = options.withFont(group.value.font);
```
Calling `withFont` with a font makes more sense.

kevinbarabash commented on src/buildMathML.js:
```diff
@@ -601,6 +601,10 @@ groupTypes.mclass = function(group, options) {
 
+// Transforms (translation/rotation) don't seem to have a representation
+// in MathML, so just treat them like \text{...}
+groupTypes.raisebox = groupTypes.text;
```
:+1:

kevinbarabash commented on src/functions.js:
```diff
@@ -51,7 +51,7 @@ defineFunction([
         type: "text",
         body: ordargument(body),
-        style: textFunctionStyles[context.funcName],
+        font: textFunctionStyles[context.funcName],
```
Since we're calling these fonts now, should probably rename this to `textFunctionFonts`.

kevinbarabash commented:
@edemaine thanks for picking this one up again and finishing it off.  The results look great!

edemaine commented:
Thanks @kevinbarabash for all the reviews today!!

ronkok commented:
I think that one can get effective presentation MathML for `\raisebox` with code something like:
````
const node = new mathMLTree.MathNode(
        "mpadded", [buildGroup(group.value.body, options)]);
node.setAttribute("voffset", group.value.dy.value);
````

I have not tested that code, but I have put the HTML below into Firefox
````
<math xmlns="http://www.w3.org/198/Math/MathML">
    <mi>a</mi><mpadded voffset="0.5em">b</mpadded><mi>c</mi>
</math>
````

... and it comes out looking like this:
![raisebox](https://user-images.githubusercontent.com/16403058/30283820-09334b7a-96ce-11e7-9fb4-a770630b1a80.PNG)

This comment is coming very late, I know. Sorry.
